### PR TITLE
Frame profiler overlay

### PR DESCRIPTION
**Pull request by hrydgard**
Initial work on a simple frame profiler overlay.

overlay

Biggest finding so far is just how slow texture decoding/loading is. It's really, really spiky in these charts, although not very visible in the screenshot above.

Turn it on by setting USE_PROFILER in profiler/profiler.h in native.

### CODE REVIEW

File: UI/GamepadEmu.cpp [hunk 1]
```diff
@@ -386,9 +386,12 @@ void InitPadLayout(float xres, float yres, float globalScale) {
...
 
-	//L and R------------------------------------------------------------
-	int l_key_X = 70 * scale;
-	int l_key_Y = 40 * scale;
+	// L and R------------------------------------------------------------
+	// Put them above the analog stick / above the buttons to the right.
+	// The corners were very hard to reach..
+
+	int l_key_X = 60 * scale;
+	int l_key_Y = yres - 110 * scale;
```
Comment (unknownbrackets) [inline UI/GamepadEmu.cpp]: Hmm, is this stuff related to the frame profiler or the other pull?

-[Unknown]

**Comment by unknownbrackets** [comment]
Seems pretty cool.  Haven't tried it yet - I don't have a good picture of which is slower (decode or upload), so interested to try it.

-[Unknown]

File: UI/GamepadEmu.cpp (hunk 1 above)
Comment (hrydgard) [inline UI/GamepadEmu.cpp]: Neither, it just ended up here by accident :P

**Comment by VIRGINKLM** [comment]
BTW, nice that you preview Tekken 6 as an example for this. Because I think it's the game with THE biggest spike of lag on PPSSPP at a specific moment, just after you load up your first fight and the first hit is about to happen the game will lag for quite long before the hit will connect and the opponent (or you) will take damage. It feels like it's related to texture loading/decoding, like that hit particle texture is REALLY heavy to load/decode.

File: UI/NativeApp.cpp [hunk 2]
```diff
@@ -48,7 +48,6 @@
 #include "file/zip_read.h"
 #include "thread/thread.h"
 #include "net/http_client.h"
-#include "gfx_es2/gl_state.h"  // should've been only for screenshot - but actually not, cleanup?
```
Comment (unknownbrackets) [inline UI/NativeApp.cpp]: This makes it so we can't compile on Windows.

-[Unknown]

**Comment by hrydgard** [comment]
@VIRGINKLM , thanks to the profiler I know what's taking time in the first hit in Tekken now :) What happens is that the lighting changes due to the hit effect, and the emulator has to compile a whole bunch new shaders, many of them complex due to Tekken using a rather complex skinning setup.

One way we could work around this in the future would be to save information about all the shaders that each game generates, and on the second startup, immediately precompile all shaders that have been used in this game before. This would cause maybe a second of startup delay, but this type of hitches would be reduced, except the first time you start up a game, of course.

**Comment by VIRGINKLM** [comment]
So an AOT shader compilation engine? That's great, I've been hoping we could take advantage in some way AOT compilation technics in games, it could help alot mobile platforms too. Makes me wonder what other can be done with AOT compilation in PPSSPP.

**Comment by unknownbrackets** [comment]
We wouldn't necessarily need a startup delay.

We could just store the list of shader ids (although then we need to version that... a different option is representative gstates), and then load them one per frame after the game starts (or even during the sceDisplay wait period.)  That would often mean not much of a stutter.

-[Unknown]

**Comment by hrydgard** [comment]
We need to change the shader generator to generate from shader ID instead of directly from gstate, which probably would be a good idea anyway..

Yeah, true, we could definitely spread it out between frames. And if we do it while the game is normally showing title screens and stuff there should barely be any impact, although I guess if the user loads a savestate they might have to live with a tiny bit of extra stuttering to spare them further stuttering later. Should be a win in any case.

Can't wait for Vulkan where we'll be able to compile shaders on multiple threads in the background...

**Comment by unknownbrackets** [comment]
Looking at GoW, I see periodic spikes in gpu loop.  I think the frame limiting may not be working well enough.  It's almost like it's letting an extra one through every 30 or 60 frames.

I also put one inside Flush, and I see it as (on the Shield TV) 40% gpu loop, 40% flush, 20% vertdec.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, GoW still has timing issues, which possibly may happen on the real hardware as well (and in that case might need some further hackery to avoid...)
